test/test_charm.py

```diff
@@ -28,50 +27,51 @@
 from ops.model import ModelError, _ModelBackend
 from ops.storage import SQLiteStorage
 
-from .test_helpers import fake_script, fake_script_calls
+from .test_helpers import FakeScriptFixture
 
+# from .test_helpers import fake_script, fake_script_calls
 
-class TestCharm(unittest.TestCase):
 
-    def setUp(self):
-        def restore_env(env: typing.Dict[str, str]):
-            os.environ.clear()
-            os.environ.update(env)
-        self.addCleanup(restore_env, os.environ.copy())
+@pytest.fixture(autouse=True)
+def env(monkeypatch: pytest.MonkeyPatch):
+    monkeypatch.setenv("PATH", str(Path(__file__).parent / 'bin'), prepend=os.pathsep)
+    monkeypatch.setenv("JUJU_UNIT_NAME", "local/0")
+    yield
 
-        os.environ['PATH'] = os.pathsep.join([
-            str(Path(__file__).parent / 'bin'),
-            os.environ['PATH']])
-        os.environ['JUJU_UNIT_NAME'] = 'local/0'
 
-        self.tmpdir = Path(tempfile.mkdtemp())
-        self.addCleanup(shutil.rmtree, str(self.tmpdir))
-        self.meta = ops.CharmMeta()
+@pytest.fixture(autouse=True)
+def events():
+    class CustomEvent(ops.EventBase):
+        pass
 
-        class CustomEvent(ops.EventBase):
-            pass
+    class TestCharmEvents(ops.CharmEvents):
+        custom = ops.EventSource(CustomEvent)
+
+    # Relations events are defined dynamically and modify the class attributes.
+    # We use a subclass temporarily to prevent these side effects from leaking.
+    ops.CharmBase.on = TestCharmEvents()  # type: ignore
+    yield
+    ops.CharmBase.on = ops.CharmEvents()  # type: ignore
 
-        class TestCharmEvents(ops.CharmEvents):
-            custom = ops.EventSource(CustomEvent)
 
-        # Relations events are defined dynamically and modify the class attributes.
-        # We use a subclass temporarily to prevent these side effects from leaking.
-        ops.CharmBase.on = TestCharmEvents()  # type: ignore
+@pytest.fixture
+def framework(request: pytest.FixtureRequest, tmp_path: pathlib.Path):
+    marker = request.node.get_closest_marker("charm_meta")  # type: ignore
```

Comment:
At first glance I really don't love this indirect way of passing data to the fixture. Are there better ways to approach it, or maybe we shouldn't use fixtures here? Maybe we can use a function (with a `request.addfinaliser`?).

Speaking of "indirect", there's also [pytest's "indirect" feature](https://stackoverflow.com/a/33879151/68707), which I don't really understand yet, but might be worth looking at.